draft-ietf-quic-tls.md

```diff
@@ -691,6 +692,24 @@ will be marked as lost before this, as they leave a gap in the sequence of
 packet numbers.
 
 
+## Discarding Initial Keys {#discard-initial}
+
+Packets protected with Initial secrets ({{initial-secrets}}) are not
+authenticated, meaning that an attacker could spoof packets with the intent to
+disrupt a connection.  To limit these attacks, Initial packet protection keys
+can be discarded more aggressively than other keys.
+
+The successful use of Handshake packets indicates that no more Initial packets
+need to be exchanged, as these keys can only be produced after receiving all
+CRYPTO frames from Initial packets.  Thus, a client MUST discard Initial keys
+when it first sends a Handshake packet and a server MUST discard Initial keys
+when it first successfully processes a Handshake packet.  Endpoints MUST NOT
+send Initial packets after this point.
```

Comment:
Should we say that this means that the packet triggering this will never be acknowledged?

Reply:
I thought the sense of the discussion in BKK was *not* to do this. It's a weird kind of implicit ACKIng. This probably needs list and issue discussion, not just merging at this time.

Reply:
Yes, and it's not clear to me it's worth making *this* change to defend against this attack. As I said, I don't recall there being consensus to make this change.

Reply:
I agree with @marten-seemann.  What this means is that implementations won't even bother sending any ACKs.  Perhaps the recommendation should be to discard the keys once the ACK arrives.

Reply:
@rpaulo To clarify, I'm saying we should not send an ACK, and we should discard the keys as soon as possible.
Delaying to discard the keys until the ACK arrives is exactly what makes the attack possible in the first place: If we rely on ACKs, there's the problem that the ACK might get lost. Then the peer has to retransmit the last Initial packet, just to trigger the ACK (although it already knows that the packet was received, because new TLS keys became available). In the end, you'll end up generating and accepting Initial packets for **a lot** longer than the 3-way handshake in which they are actually needed, which is a big window for an attacker to inject a packet that will kill the connection.

Reply:
Okay, I see what you mean.  I think that from a transport protocol perspective, not sending an ACK isn't good since you just ignored a possible RTT estimation.  You could estimate the RTT from the new handshake packet, I suppose, but that's not how the transport part works.

I'm not sure I understand the attack since to spoof the initial packet,  the third party must be doing a man in the middle attack.  In that case, why not do it early and abuse the stateless reset or retry?

Reply:
It is not a man in the middle attack, it is a man on the side attack. The initial packet can be decrypted and a response can be constructed, but at the next encryption level that option is gone. This is why the initial is special. 

If it is possible to forge a reset from the initial packet, something is wrong in the design. It is not possible when a new connection id is produced, but I'm not sure if there the CID is fixed zero length. If there is an attack here, the reset token ought to moved further into the handshake.

Reply:
I agree with the spirit of the text: we must defend against this attack, and the proper way has to be some form of implicit ACK. But I would like to see the implicit acknowledgement mechanism specified in the transport draft, because it affects the transport state machine. Obviously, bad things would happen if one side decided to stop send sending Initial packets while the other side was still waiting for ACK of its own packets, and repeating them. That means we must be crystal clear on when exactly to perform this "initial cut-off": on the client side, when the Handshake secret is successfully received and the client starts being able receiving handshake packets; on the server side, when the first handshake packet is received from the server.

Reply:
I also looked at how to implement a similar protection without this kind of cut-off, and it just does not work. Suppose for example that the client receives a repeated Initial packet after the cut-off. In the post cutoff defensive mode, it can decide to ignore the packet content, since it cannot possibly learn anything interesting, but it would still need to send an ACK, otherwise the peer will keep repeating 3 or 4 times and then give up on the connection. But then, if the packet from the "server" was forged, the ACK from the client will be treated as a protocol violation, and the server will close the connection. Catch 22, no good way to get out of that.